spec/compiler/semantic/instance_var_spec.cr

```diff
@@ -5387,4 +5387,166 @@ describe "Semantic: instance var" do
       ),
       "can't infer the type of instance variable '@x' of Foo"
   end
+
+  it "errors when overriding inherited instance variable with incompatible type" do
+    assert_error <<-CR, "instance variable '@a' of A must be Int32, not (Char | Int32)"
+      class A
+        @a = 1
+      end
+
+      class B < A
+        @a = 'a'
+      end
+      CR
+  end
+
+  it "accepts overriding inherited instance variable with compatible type" do
+    semantic <<-CR
+      class A
+        @a = 1
+      end
+
+      class B < A
+        @a = 2
+      end
+      CR
+  end
+
+  describe "instance variable inherited from multiple parents" do
+    context "with compatible type" do
+      it "module and class, with declarations" do
+        assert_error <<-CR, "instance variable '@a' of B is already defined in A"
+          module M
+            @a : Int32 = 1
+          end
+
+          class A
+            @a : Int32 = 2
+          end
+
+          class B < A
+            include M
+          end
+          CR
+      end
+
+      it "module and class, with definitions" do
+        assert_error <<-CR, "instance variable '@a' of B is already defined in A"
+          module M
+            @a = 1
+          end
+
+          class A
+            @a = 2
+          end
+
+          class B < A
+            include M
+          end
+          CR
+      end
+
+      it "accepts module and module, with definitions" do
+        semantic <<-CR
+          module M
+            @a = 1
+          end
+
+          module N
+            @a = 2
+          end
+
+          class B
+            include N
+            include M
+          end
+          CR
+      end
+
+      it "accepts module and module, with declarations" do
+        semantic <<-CR
+          module M
+            @a : Int32?
+          end
+
+          module N
+            @a : Int32?
+          end
+
+          class B
+            include N
+            include M
+          end
+          CR
+      end
+    end
+
+    context "with incompatible type" do
+      it "module and class, with definitions" do
+        assert_error <<-CR, "instance variable '@a' of B is already defined in A"
+          module M
+            @a = 'a'
+          end
+
+          class A
+            @a = 1
+          end
+
+          class B < A
+            include M
+          end
+          CR
+      end
+
+      it "module and class, with declarations" do
+        assert_error <<-CR, "instance variable '@a' of B is already defined in A"
```

Comment:
If you use `M` as a shard and you don't have control over it, what are you options to move forward?

Reply:
At the moment, none, but in the future we might try to be smarter. Ideally for me, it should allow shadowing with scoping, but this is not something to discuss here, right?

Reply:
There are many things you can't do anything about when you don't have control over the code 🤷